clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,340 @@
+                    Collections.unmodifiableMap(new LinkedHashMap<>(sessionParts));
+                return new FetchRequestData(toSend, Collections.<TopicPartition>emptyList(), toSend, nextMetadata);
+            }
+            int numAdded = 0, numRemoved = 0, numAltered = 0;
+            List<TopicPartition> toForget = new ArrayList<>(0);
+            for (Iterator<Entry<TopicPartition, PartitionData>> iter =
+                     sessionParts.entrySet().iterator(); iter.hasNext(); ) {
+                Entry<TopicPartition, PartitionData> entry = iter.next();
+                TopicPartition topicPartition = entry.getKey();
+                PartitionData prevData = entry.getValue();
+                PartitionData nextData = next.get(topicPartition);
+                if (nextData != null) {
+                    if (prevData.equals(nextData)) {
+                        // Omit this partition from the FetchRequest, because it hasn't changed
+                        // since the previous request.
+                        next.remove(topicPartition);
+                    } else {
+                        // Move the altered partition to the end of 'next'
+                        next.remove(topicPartition);
+                        next.put(topicPartition, nextData);
+                        entry.setValue(nextData);
+                        numAltered++;
+                    }
+                } else {
+                    // Set maxBytes to 0 to indicate that we no longer want to listen to this partition.
+                    next.remove(topicPartition);
+                    toForget.add(topicPartition);
+                    // Remove this partition from the session.
+                    iter.remove();
+                    numRemoved++;
+                }
+            }
+            // Add any new partitions to the session.
+            for (Iterator<Entry<TopicPartition, PartitionData>> iter =
+                     next.entrySet().iterator(); iter.hasNext(); ) {
+                Entry<TopicPartition, PartitionData> entry = iter.next();
+                TopicPartition topicPartition = entry.getKey();
+                PartitionData nextData = entry.getValue();
+                if (sessionParts.containsKey(topicPartition)) {
+                    // In the previous loop, all the partitions which existed in both sessionParts
+                    // and next were moved to the end of next, or removed from next.  Therefore,
+                    // once we hit one of them, we know there are no more unseen entries to look
+                    // at in next.
+                    break;
+                }
+                sessionParts.put(topicPartition, nextData);
+                numAdded++;
+            }
+            log.debug("Built incremental fetch {} for node {}. numAdded = {}, numRemoved = {}, " +
+                "numAltered = {}, sessionSize = {}", nextMetadata, node, numAdded, numRemoved,
+                numAltered, sessionParts.size());
```

Comment:
Same as above. It will probably be particularly useful for incremental fetches to have the partitions explicitly in the log message.